rapid7/meterpreter
# support building on newer Linux systems and Makefile cleanups

This fixes some issues building Meterpreter under recent linux distributions and quiets some of the noise to make further development feasible.

 - try to share some bits between different makefiles, make modifying global compiler flags not such a huge pain.
 - directly specify we should be using the gold rather than bpf linker
 - make compiler output largely quiet except where we care - allow warnings to actually be visible
 - don't delete downloaded tarballs with --really-clean
 - add missing dependencies between libraries (--no-add-needed/--no-copy-dt-needed-entries causes lots of trouble)
 - update readme to show what to install to build

I made minimal changes to the loader makefile - it breaks easily.
 -Os prevents if from being able to load libc, for instance

# Verification Steps

On Ubuntu 14.04 or Fedora 21 x64, try these things:
- [x] install recommended packages in the README.md file
- [x] run 'make really-clean && make && make install'
- [x] verify that it runs and installs new binaries in your metasploit-framework directory
- [x] verify that you can start a meterpreter shell
- [x] load networkpug
- [x] load sniffer

Build should look a little nicer, it stashes a lot of the 3rd party build noise into log files:
```
$ make
Building libc
Building libm
Building libdl
Building OpenSSL
Installing libcrypto
Installing libssl
make -C workspace/common
make[1]: Entering directory `/home/bcook/projects/meterpreter/workspace/common'
[CC] args.o
[CC] base.o
[CC] unix_socket_server.o
[CC] passfd_server.o
[CC] ptrace.o
[CC] base_inject.o
[CC] base_dispatch.o
[CC] base_dispatch_common.o
[CC] buffer.o
[CC] channel.o
[CC] common.o
```

Though the extensive use of 'HANDLE' in a unix environment is imperfect, so there are lots of bad casts and things that will be solved later. My editor also auto-deletes trailing whitespace on save, so sorry in advance for the noise.
